File: docking/docking/image_analyzer.py
import cv2
import math
import time
import numpy as np
from apriltag import apriltag

TAG16               = "tag16h5" # tag family 
TAG36               = "tag36h11"
MIN_MARGIN          = 10 # Filter value for tag detection
detector            = apriltag(TAG36)
CAMERA_HORIZ_FOV    = 62.2
CAMERA_VERTICAL_FOV = 48.8
CENTRAL_TAG_SIZE    = 0.227 # size of one side of tag in m
PERIPHERAL_TAG_SIZE = 0.045

class ImageAnalyzer:

    def process_image(self, img, ind, yaw):
        """
        Process an image to determine how far off the drone is
        Return the x, y, and z offset in meters, yaw offset in degrees clockwise
        ind tells simulator which tag to target (0 for center, 1-8 counterclockwise from North for perimeter)
        yaw is the drone's current rotation, used for rectifying the rotation back to north before processing
        """

        # get image dimensions
        height, width, _ = img.shape
        img_center = (int(width / 2), int(height / 2))

        # rotate to due north, so we can use north/east/down (ned) coordinates instead of body coordinates
        image_center = tuple(np.array(img.shape[1::-1]) / 2)
        rot_mat = cv2.getRotationMatrix2D(image_center, yaw * -1, 1.0)
        result = cv2.warpAffine(img, rot_mat, img.shape[1::-1], flags=cv2.INTER_LINEAR, borderValue=(255, 255, 255))
        img = result

        # detect
        greys = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
        dets = detector.detect(greys)
        
        tags_detected=""
        for det in dets:
            if(det["margin"] >= MIN_MARGIN):
                tags_detected += str(det["id"]) + " "

        for det in dets:
            if det["id"] == ind:
                if det["margin"] >= MIN_MARGIN:
                    det_center = det["center"].astype(int)

                    # Extract apriltag bounding box
                    rect = det["lb-rb-rt-lt"].astype(int).reshape((-1,1,2))

                    # Calculates average side length in pixels
                    sidelist = list()
                    for i in range (0,4):
                        for j in range (i+1,4):
                            sidelist.append(math.sqrt((rect[i][0][0] - rect[j][0][0]) ** 2 + (rect[i][0][1] - rect[j][0][1]) ** 2))
                    sidelist.sort()
                    sideAvg = 0
                    for i in range(4):
                        sideAvg = sideAvg+sidelist[i]
                    sideAvg = sideAvg/4.0
                    tag_pixel_ratio = (CENTRAL_TAG_SIZE if ind == 0 else PERIPHERAL_TAG_SIZE) / sideAvg # ratio of meters to pixels

                    # Absolute difference in height (meters)
                    alt_err = 0.50 / math.tan(math.radians(CAMERA_HORIZ_FOV * 0.50)) * width * tag_pixel_ratio

                    # Finds difference in rotation
                    y3 = rect[3][0][1]
                    y0 = rect[0][0][1]
                    x3 = rect[3][0][0]
                    x0 = rect[0][0][0]

                    if x3 - x0 != 0:
                        # Look at bottom left corner and top left corner to determine angle from y-axis
                        incline_angle = math.degrees(math.atan2(y3 - y0, x3 - x0))

                        # Convert angle so that 0 degrees is north, positive angles are clockwise, negative angles are ccw
                        # -180 < incline_angle < 180
                        incline_angle = incline_angle * -1 + 90
                        if incline_angle > 0:
                            incline_angle = incline_angle - 180
                        else:
                            incline_angle = incline_angle + 180
                    elif y3 > y0:
                        incline_angle = -180
                    else:
                        incline_angle = 0
                    rot_err = incline_angle * -1 # to make this cw

                    # Absolute horizontal difference (meters)
                    x_offset = det_center[0] - img_center[0] # offset east from center
                    y_offset = img_center[1] - det_center[1] # offset north from center
                    x_err = x_offset * tag_pixel_ratio
                    y_err = y_offset * tag_pixel_ratio

                    # cv2.polylines(img, [rect], True, (0, 0, 255), 2)
                    # cv2.imshow("image analyzed", img)
                    # cv2.waitKey(1)

                    return float(x_err), float(y_err), float(alt_err), float(rot_err), tags_detected

                    break

        print("Desired apriltag not found, aborting")
        return

        # cv2.imshow("TITLE", img)
        # cv2.waitKey()

# if __name__ == "__main__":
#     img = cv2.imread("/home/axel/Pictures/a5.png")

#     scale_percent = 30 # percent of original size
#     width = int(img.shape[1] * scale_percent / 100)
#     height = int(img.shape[0] * scale_percent / 100)
#     dim = (width, height)
#     # resize image
#     resized = cv2.resize(img, dim, interpolation = cv2.INTER_AREA)

#     image_analyzer = ImageAnalyzer()
#     x_err, y_err, z_err, rot_err, e = image_analyzer.process_image(resized, 1)

File: docking/cpp_docking/src/image_analyzer.cpp
#include "image_analyzer.hpp"
#include "util.hpp"

#include <opencv2/core.hpp>
#include <opencv2/core/types.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>

#include <apriltag/apriltag.h>
#include <apriltag/tag36h11.h>

#include <string>
#include <math.h>
#include <vector>
#include <algorithm>

ImageAnalyzer::ImageAnalyzer(){
    tf=tag36h11_create();
    m_tagDetector=apriltag_detector_create();
    apriltag_detector_add_family(m_tagDetector,tf);

}
ImageAnalyzer::~ImageAnalyzer(){
    apriltag_detector_destroy(m_tagDetector);
    tag36h11_destroy(tf);
}

float* ImageAnalyzer::processImage(Mat img, int ind, float yaw, std::string& tags){
    float* errs=new float[4];

    int width=img.cols;
    int height=img.rows;
    Point2f image_center(img.cols / 2, img.rows / 2);

    //Rotates target image and converts to greyscale
    Mat rot_mat = getRotationMatrix2D(image_center, -1.0*yaw, 1.0);
    warpAffine(img, img, rot_mat, Size(img.cols, img.rows), INTER_LINEAR, BORDER_CONSTANT, Scalar(255, 255, 255));
    cvtColor(img,img,COLOR_BGR2GRAY);

    image_u8_t im={.width=img.cols,.height=img.rows, .stride=img.cols, .buf=img.data}; //Converts CV2 image to C-friendly image format

    zarray_t* dets=apriltag_detector_detect(m_tagDetector,&im); //Detects apriltags

    //Prints detected tags
    std::string tagsDetected="";
    for(int i=0;i<zarray_size(dets);i++){
        apriltag_detection_t *det;
        zarray_get(dets,i,&det);
        if(det->decision_margin>=MARGIN){
            tagsDetected+=std::to_string(det->id)+" ";
        }
    }
    tags=tagsDetected;

    //Iterates over detected tag to find target tag
    for(int i=0;i<zarray_size(dets);i++){
        apriltag_detection_t *det;
        zarray_get(dets,i,&det);

        if(det->id==ind&&det->decision_margin>=MARGIN){ //Target tag found

            //Gathers location information on detection
            double* center=det->c;
            double rect[4][2];
            for(int i=0;i<4;i++){
                for(int j=0;j<2;j++){
                    rect[i][j]=(det->p)[i][j];
                }
            }

            //Calculates average side length
            std::vector<float> sideList;
            for(int i=0;i<4;i++){
                for(int j=i+1;j<4;j++){
                    sideList.push_back((float)sqrt(pow((rect[i][0]-rect[j][0]),2)+pow((rect[i][1]-rect[j][1]),2)));
                }
            }
            std::sort(sideList.begin(),sideList.end());
            float sideAvg=0;
            for(int i=0;i<4;i++){
                sideAvg+=sideList[i];
            }
            sideAvg/=4.0;

            //Calculates pixel density of image
            float tag_pixel_ratio;
            if(ind==0){
                tag_pixel_ratio=((float)CENTRAL_TAG_SIZE)/sideAvg;
            }
            else{
                tag_pixel_ratio=((float)PERIPHERAL_TAG_SIZE)/sideAvg;
            }

            errs[2]=0.50/tan((CAMERA_FOV_HORIZONTAL*0.50)*M_PI/180.0)*width*tag_pixel_ratio; //Calculates alt_err

            float y3=rect[3][1]; //Chooses points for calculating diagonal angle
            float y0=rect[0][1];
            float x3=rect[3][0];
            float x0=rect[0][0];

            //Calculates diagonal angle
            float incline_angle;
            if(x3-x0!=0){
                incline_angle=180.0/M_PI *atan2(y3-y0,x3-x0);
                incline_angle=incline_angle*-1.0+90.0;
                if(incline_angle>0){
                    incline_angle-=180;
                }
                else{
                    incline_angle+=180;
                }
            }
            else if(y3>y0){
                incline_angle=-180;
            }
            else{
                incline_angle=0;
            }
            errs[3]=incline_angle*-1; //Finds rotational error from diagonal angle

            //Finds the offset of the image location, finds x/y err
            float x_offset=center[0]-image_center.x;
            float y_offset=image_center.y-center[1];
            errs[0]=x_offset*tag_pixel_ratio;
            errs[1]=y_offset*tag_pixel_ratio;

            //Cleanup
            apriltag_detection_destroy(det);
            zarray_destroy(dets);
            return errs;
        }
        apriltag_detection_destroy(det);
    }
    zarray_destroy(dets);
    return nullptr; //Tag not detected, returns nullptr
}